src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -413,330 +318,193 @@ class DataBox<tmpl::list<Tags...>>
                       tmpl::bind<detail::first_matching_tag,
-                                 tmpl::pin<tmpl::list<Tags...>>, tmpl::_1>>{});
-  add_sub_compute_item_tags_to_box<Tag>(
-      typename Subitems<Tag>::type{},
-      typename detail::has_return_type_member<Subitems<Tag>>::type{});
+                                 tmpl::pin<TagsList>, tmpl::_1>>{});
 }
-// End adding compute items
+}  // namespace detail
 
-// Adding simple items
 template <typename... Tags>
-template <typename ParentTag, typename... Subtags>
+template <typename ParentTag, typename... SubitemTags>
 SPECTRE_ALWAYS_INLINE constexpr void
-db::DataBox<tmpl::list<Tags...>>::add_subitem_tags_to_box(
-    tmpl::list<Subtags...> /*meta*/) noexcept {
-  const auto helper = [this](auto tag_v) {
+db::DataBox<tmpl::list<Tags...>>::add_mutable_subitems_to_box(
+    tmpl::list<SubitemTags...> /*meta*/) noexcept {
+  const auto add_mutable_subitem_to_box = [this](auto tag_v) {
     (void)this;  // Compiler bug warns this is unused
-    using tag = decltype(tag_v);
-    using ItemType = detail::storage_type<tag, tmpl::list<Tags...>>;
-    get_deferred<tag>() = Deferred<ItemType>(ItemType{});
-    Subitems<ParentTag>::template create_item<tag>(
-        make_not_null(&get_deferred<ParentTag>().mutate()),
-        make_not_null(&get_deferred<tag>().mutate()));
+    using subitem_tag = decltype(tag_v);
+    get_item<subitem_tag>() =
+        detail::Item<subitem_tag>(typename subitem_tag::type{});
+    Subitems<ParentTag>::template create_item<subitem_tag>(
+        make_not_null(&get_item<ParentTag>().mutate()),
+        make_not_null(&get_item<subitem_tag>().mutate()));
   };
 
-  EXPAND_PACK_LEFT_TO_RIGHT(helper(Subtags{}));
+  EXPAND_PACK_LEFT_TO_RIGHT(add_mutable_subitem_to_box(SubitemTags{}));
 }
 
 template <typename... Tags>
-template <size_t ArgsIndex, typename Tag, typename... Ts>
+template <size_t ArgsIndex, typename MutableItemTag, typename... Ts>
 SPECTRE_ALWAYS_INLINE constexpr char
-db::DataBox<tmpl::list<Tags...>>::add_item_to_box(
-    std::tuple<Ts...>& tupull) noexcept {
+db::DataBox<tmpl::list<Tags...>>::add_mutable_item_to_box(
+    std::tuple<Ts...>& items) noexcept {
   using ArgType = std::tuple_element_t<ArgsIndex, std::tuple<Ts...>>;
-  static_assert(not tt::is_a<Deferred, std::decay_t<ArgType>>::value,
-                "Cannot pass a Deferred into the DataBox as an Item. This "
-                "functionally can trivially be added, however it is "
-                "intentionally omitted because users of DataBox are not "
-                "supposed to deal with Deferred.");
-  get_deferred<Tag>() =
-      Deferred<detail::storage_type<Tag, tmpl::list<Tags...>>>(
-          std::forward<ArgType>(std::get<ArgsIndex>(tupull)));
-  add_subitem_tags_to_box<Tag>(typename Subitems<Tag>::type{});
+  get_item<MutableItemTag>() = detail::Item<MutableItemTag>(
+      std::forward<ArgType>(std::get<ArgsIndex>(items)));
```

Comment:
Do you think we might be able to relax the requirement that the exact same type that the tag holds must be passed in to `db::create`? I think it should be enough that the passed-in type is _move-assignable_ to the type that the tag holds.

Reply:
do you have an example in mind?

Reply:
This could be possible, but I worry it'll make the construction and resulting errors nearly impossible to understand. Changing `std::is_same_v` to `std::is_convertible_v` would allow implicit conversions, which might be okay. `is_constructible_v` could be used to check explicit construction. I'm also not clear of the benefits of adding this as an internal feature rather than doing an explicit conversion where you need it.